### PR TITLE
feat(query processor) Parse conditions

### PR DESCRIPTION
**Pull request by fpacifici**
This takes the same approach of complex_column_expr to preserve the current parser logic. It takes conditions_expr (that essentially defines lexing and parsing for conditions), it makes that function parametric so that it can be used to build any type of expressions.

It also provides a function `parse_conditions_to_expr` that parses the Snuba conditions into a FunctionCall. 

Limitation: it does not resolve columns yet. This is not a blocker, it will be useful later when processing joins.

test plan:
- added a test to mimic test_util, but for the new data model

### CODE REVIEW

File: snuba/query/parser/conditions.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+
+TExpression = TypeVar("TExpression")
+
+
+class InvalidConditionException(Exception):
+    pass
+
+
+def parse_conditions(
```
Comment (fpacifici) [inline snuba/query/parser/conditions.py]: This is essentially a cut/paste of the current algorithm, but it is parametric. So it preserves all the corner cases.

File: tests/query/parser/test_conditions.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+from snuba.query.conditions import (
+    ConditionFunctions,
+    BooleanFunctions,
+)
+from snuba.query.parser.conditions import parse_conditions_to_expr
+from snuba.util import tuplify
+
+test_conditions = [
+    ([], None,),
+    ([[[]], []], None,),
```
Comment (fpacifici) [inline tests/query/parser/test_conditions.py]: I think this should be a parse error, but apparently today it is supported so I am preserving the behavior.

File: snuba/query/parser/conditions.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                depth + 1,
+            )
+            for cond in conditions
+        )
+        return or_builder([s for s in sub if s])
+    else:
+        raise InvalidConditionException(str(conditions))
+
+
+def parse_conditions_to_expr(
```
Comment (fpacifici) [inline snuba/query/parser/conditions.py]: See the test case to have an idea of how it works and what it produces.

File: snuba/query/parser/conditions.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        return binary_condition(
+            None,
+            function,
+            expressions[0],
+            multi_expression_builder(expressions[1:], function),
+        )
+
+    def and_builder(expressions: Sequence[Expression]) -> Optional[Expression]:
+        if not expressions:
+            return None
```
Comment (fpacifici) [inline snuba/query/parser/conditions.py]: This is for the type checker to keep the return type of multi_expression_builder non nullable which helps with the recursion.

File: snuba/query/columns.py [hunk 5]
```diff
@@ -2,19 +2,19 @@
 import re
 
 from datetime import date, datetime
-from typing import Any, List, Optional, OrderedDict, Tuple, Union
+from typing import Any, List, Optional, Sequence, Tuple, Union
 import _strptime  # NOQA fixes _strptime deferred import issue
 
 from snuba.clickhouse.escaping import escape_alias, NEGATE_RE
 from snuba.query.parser.functions import parse_function
+from snuba.query.parser.conditions import parse_conditions
```
Comment (tkaemming) [inline snuba/query/columns.py]: nit: Sort order should be swapped here.

File: snuba/query/parser/conditions.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+      left hand side, an operator and a literal on the right hand side.
+    """
+    from snuba.clickhouse.columns import Array
+
+    if not conditions:
+        return None
+
+    if depth == 0:
+        # dedupe conditions at top level, but keep them in order
+        sub = OrderedDict(
```
Comment (tkaemming) [inline snuba/query/parser/conditions.py]:
```
snuba/query/parser/conditions.py:61: error: "TypeAlias" not callable
```

This is now new, but I think this is indicating that this should be using `collections.OrderedDict` instead of the generic variant.

**Review by tkaemming — APPROVED** [review]
This is super cool.

File: tests/query/parser/test_conditions.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+                ),
+            ),
+        ),
+    ),
+    (
+        [[["a", "=", 1], []]],
+        FunctionCall(
+            None, ConditionFunctions.EQ, (Column(None, "a", None), Literal(None, 1)),
+        ),
+    ),  # Malformed Condition Input
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply tests/query/parser/test_conditions.py]: I think we should, but it seems that now our syntax is permissive around that so we should ensure this is not happening in production. This test came from the existing test_util.py

File: tests/query/parser/test_conditions.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+        ),
+    ),  # Test functions in conditions
+    (
+        [["primary_hash", "LIKE", "%foo%"]],
+        FunctionCall(
+            None,
+            ConditionFunctions.LIKE,
+            (Column(None, "primary_hash", None), Literal(None, "%foo%")),
+        ),
+    ),  # Test special output format of LIKE
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply tests/query/parser/test_conditions.py]: I think this was `special` because of the escaping maybe ? 
I copied from here
https://github.com/getsentry/snuba/blob/master/tests/test_util.py#L129-L134

Does not seem very special to me either.

File: tests/query/parser/test_conditions.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply tests/query/parser/test_conditions.py]: That makes sense to me — just making sure this one wasn't an intended long-term behavior since the other one was called out specifically above.
